### PR TITLE
[DependencyInjection] Add support for named arguments

### PR DESCRIPTION
**Pull request by dunglas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | todo

This PR introduces named arguments for services definitions. It's especially useful to inject parameters in an autowired service. It is (at least partially) an alternative to #21376 and #20738.

Usage:

```yml
services:
    _defaults: { autowire: true }
    Acme\NewsletterManager: { $apiKey: "%mandrill_api_key%" }

# Alternative (traditional) syntax
services:
    newsletter_manager:
        class: Acme\NewsletterManager
        arguments:
            $apiKey: "%mandrill_api_key%"
        autowire: true
```
```php
use Doctrine\ORM\EntityManager;
use Psr\Log\LoggerInterface;

namespace Acme;

class NewsletterManager
{
    private $logger;
    private $em;
    private $apiKey;

    public function __construct(LoggerInterface $logger, EntityManager $em, $apiKey)
    {
        $this->logger = $logger;
        $this->em = $em;
        $this->apiKey = $apiKey;
    }
}
```

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
I like it.
Should be made to work with method calls.
Missing resources tracking for now.
With now two alts to #20738, I think we should close it :)

**Comment by fabpot** [comment]
Note that this conflicts with the implementation of #21313 that was just merged.

**Comment by ogizanagi** [comment]
> Note that this conflicts with the implementation of #21313 that was just merged.

Not really (AFAIU). But can't be used along with.

Side-note: I've already seen some projects using arguments keys, despite it has no effect (simply for "increasing readability"). Shouldn't it be considered as a BC break?

**Comment by fabpot** [comment]
```yaml
services:
    App\Foo\Bar: {'@baz', 'foo', apiKey: "%mandril_api_key%"}
```

**Comment by dunglas** [comment]
@ogizanagi in my opinion no, it was explicitly documented to support only packed arrays (see my changes in `Definition`).
If it really matters, I can skip named parameters not matching instead of throwing, but it will be harder to debug.

**Comment by dunglas** [comment]
> Not really (AFAIU). But can't be used along with.

Indeed, I'll try to adapt my PR to be able to use both syntaxes together.

This would be nice but it actually doesn't work:
```
services:
    Acme\NewsletterManager: { apiKey: "%mandril_api_key%" }
```

But this works:


```
services:
    Acme\NewsletterManager:
        arguments:
            apiKey: "%mandril_api_key%"
```

**Comment by dunglas** [comment]
@nicolas-grekas 

> Should be made to work with method calls.

It would be a nice addition, I propose to do it in another PR when this one will be reviewed and merged.

> Missing resources tracking for now.
Done, I rely on @weaverryan's implementation for now because conditions to rebuild the container will be the same if we add `calls` support later.

> With now two alts to #20738, I think we should close it :)

Done!

**Comment by kasparsklavins** [comment]
Should there really exist 101 different ways to define the same service?

**Comment by dunglas** [comment]
@kasparsklavins we are trying to find the best way to inject parameters when using autowiring and service auto-registration (#21289). It will be a brand new way to use Symfony that will change drastically the first experience for new users and allow develop faster.
For now, there is no way to do it. Please consider the big picture. Btw, you can already define a service using XML, YAML, PHP your own loader, autowiring and so on...

**Comment by dunglas** [comment]
@fabpot @ogizanagi I've tried something to support the short syntax: https://github.com/symfony/symfony/pull/21383/commits/b8c0e7cff89818122e7071b5c305c521e002bb61 WDYT?

**Comment by ogizanagi** [comment]
👍 Looks good to me, as soon as we warn in the documentation no argument should be named the same as one of the definition keywords, otherwise it won't work. 
I don't know if we can do better for now :)

**Comment by ogizanagi** [comment]
Re-thinking about it, it also means a introducing a new keyword someday means a potential BC break for someone using the short syntax along with named arguments 😕 
So maybe this tradeoff isn't worth it and the short syntax should be kept for most simple usages.

**Comment by nicolas-grekas** [comment]
Enforce a leading `$`, eg `$apiKey: '%foo%'`?
About method calls, is it that a big change ? Because that feels incomplete to me to not have it right now...

**Comment by iltar** [comment]
One minor issue I see with this: when you rename a variable in php, it will now crash, but this won't be detected with applications (or bundles) with only unit-tests.

**Comment by dunglas** [comment]
@iltar you're right but it's already the case for traditional definitions (if you add or remove an argument it will fail) and for `calls` definitions (if you rename the method it will fail).

It's why unit testing is not enough and apps need to have a pyramid of tests including some functional and UI ones.

**Comment by iltar** [comment]
@dunglas correct, however, there's a bunch of refactoring tools and find/replaces that might modify the name of a variable. Correctly removing one without causing parse errors is a lot more complex, hence this issue is faster to present itself.

**Comment by dunglas** [comment]
PR updated to make named arguments starting with a `$` as suggested by @nicolas-grekas. It solves problems raided by @ogizanagi (no conflict possible now) and partially the one solved by @iltar (the $ makes it explicit that there is a special handling, and IDE can add support for this new convention to allow fluent refactoring).

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+                    continue 2;
+                }
+            }
+
+            throw new InvalidArgumentException(sprintf('Unable to resolve the argument "%s" of the service "%s": the constructor of the class "%s" has no argument of this name.', $key, $this->currentId, $class));
+        }
+
+        if ($originalArguments !== $arguments) {
+            $value->setArguments($arguments);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php]: `ksort($arguments)` before set?

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php [hunk 2]
```diff
@@ -227,6 +227,22 @@ private function parseDefaults(array &$content, $file)
...
 
     /**
+     * @param array $service
+     *
+     * @return bool
+     */
+    private function isUsingShortSyntax(array $service)
+    {
+        foreach ($service as $key => $value) {
+            if (!is_int($key) && '$' !== substr($key, 0, 1)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: should be an `||` instead of `&&`?
I'd write it as `if (!isset($key[0]) || '$' !== $key[0]) {`

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php [hunk 3]
```diff
@@ -245,7 +261,7 @@ private function parseDefinition($id, $service, $file, array $defaults)
             return;
         }
 
-        if (is_array($service) && array_values($service) === $service) {
+        if (is_array($service) && $this->isUsingShortSyntax($service)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: note that array_values checked the order of keys, whereas isUsingShortSyntax doesn't anymore, isn't it?

Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: Indeed, is it a problem?

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php (hunk 2 above)
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: > should be an || instead of &&?

No it's correct (only int and keys starting with a $ are allowed when using short syntax)

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php [hunk 4]
```diff
@@ -227,6 +227,22 @@ private function parseDefaults(array &$content, $file)
...
 
     /**
+     * @param array $service
+     *
+     * @return bool
+     */
+    private function isUsingShortSyntax(array $service)
+    {
+        foreach ($service as $key => $value) {
+            if (!is_int($key) && (isset($key[0]) && '$' !== $key[0])) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: the empty string case should also return false, isn'it? thus:
`if (is_string($key) && ('' === $key || '$' !== $key[0])) {`

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php (hunk 3 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: no :)

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php (hunk 4 above)
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: Why? If it's an empty, string, it's not a short definition.

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: that's exactly what I'm saying :)

Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: done

**Comment by javiereguiluz** [comment]
I don't like the `$` at all. It makes the YAML file look a mix of PHP and YAML contents. Example:

```yaml
arguments:
    $apiKey: "%mandrill_api_key%"
```

**Comment by zanderbaldwin** [comment]
I think the `$` is appropriate. It makes it much more obvious that the keys are directly referencing PHP variables/parameters (as defined in the method definition) rather than some non-trivial structure defined by something like the config component.
Making the arguments configuration seem a little less "magic" to newer developers (meaning a lower learning curve) is, to me personally, more important than clean looking configuration files.

File: src/Symfony/Component/DependencyInjection/Definition.php [hunk 5]
```diff
@@ -189,17 +189,17 @@ public function addArgument($argument)
...
      *
      * @return $this
      *
      * @throws OutOfBoundsException When the replaced argument does not exist
      */
     public function replaceArgument($index, $argument)
     {
-        if ($index < 0 || $index > count($this->arguments) - 1) {
-            throw new OutOfBoundsException(sprintf('The index "%d" is not in the range [0, %d].', $index, count($this->arguments) - 1));
+        if (!array_key_exists($index, $this->arguments)) {
```
Comment (GuilhemN) [inline src/Symfony/Component/DependencyInjection/Definition.php]: Shouldn't we throw an exception if the key isn't prefixed with `$`?

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    {
+        if (!$value instanceof Definition) {
+            return parent::processValue($value, $isRoot);
+        }
+
+        $class = $value->getClass();
+        $originalArguments = $arguments = $value->getArguments();
+
+        foreach ($arguments as $key => $argument) {
+            if (!is_string($key)) {
```
Comment (chalasr) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php]: Should be `!is_string($key) || '$' !== $key[0]`, right? Given `<argument index="0">foo</argument>` would reach the exception below

Comment (GuilhemN) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php]: `foo` should be forbidden imo so I think an exception should be thrown.
However, https://github.com/symfony/symfony/pull/21383/files#r97614352 seems to be a better place for that.

File: src/Symfony/Component/DependencyInjection/Definition.php (hunk 5 above)
Comment (chalasr) [inline src/Symfony/Component/DependencyInjection/Definition.php]: It would be a BC break since indexed arguments are valid (i.e. `replaceArgument(0, $arg)`).

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php (hunk 6 above)
Comment (chalasr) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php]: We are talking about the key being `0`, `foo` being the value. The example I given must stay valid and would actually fail here.

File: src/Symfony/Component/DependencyInjection/Definition.php (hunk 5 above)
Comment (GuilhemN) [inline src/Symfony/Component/DependencyInjection/Definition.php]: I meant for string indexes :)

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php (hunk 6 above)
Comment (GuilhemN) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php]: Indeed you're right, should be updated :+1: (at least for `index_` arguments).

File: src/Symfony/Component/DependencyInjection/Definition.php (hunk 5 above)
Comment (chalasr) [inline src/Symfony/Component/DependencyInjection/Definition.php]: Ah indeed 👍

**Comment by nicolas-grekas** [comment]
:+1:

**Comment by weaverryan** [comment]
Hi guys!

Ok, I like this approach! Obviously, this should only be used by app code, and it's a really sensible way to fill in the "gap" of autowiring with scalar values.

But, because it's a bit of magic, let's make sure it behaves as predictably as possible! Here are some edge cases I've found, which are mostly "mistakes" that yield confusing errors. Btw, if you do *not* enable autowiring on a service (which is legal, but likely an edge case, or maybe an accident!), when you make a mistake, the errors are usually worse.

A) I forget to map an argument:

```php
    public function __construct($arg1, $arg2)
    {
        $this->arg1 = $arg1;
        $this->arg2 = $arg2;
    }
```

```yml
services:
    fun_generator:
        class: AppBundle\Services\FunGenerator
        arguments:
            $arg2: 'foo_bar'
```

Without autowiring, the error is:

> Missing argument 2 for AppBundle\Services\FunGenerator::__construct() ... (from the cached container)

Basically, because arg1 was never mapped, when the container is dumped (or the service is created from the Definition), arg2 becomes the "first" argument.

Instead, ideally, we would say something like (before dumping):

> Missing argument 1 for AppBundle\Services\FunGenerator::__construct() in your service configuration.

This is tricky... because `arg1` might still be autowired (if it had a type-hint)... or... in theory... filled in by some other compiler pass. But allowing `$arg2` to be ultimately passed as `$arg1` is just wrong.

With autowiring, the same situation gives this error

> Cannot autowire service "fun_generator": argument $arg1 of method AppBundle\Services\FunGenerator::__construct() must have a type-hint or be given a value explicitly.

Which seems quite good to me :).

B) I forget the $ prefix!

```yml
services:
    fun_generator:
        class: AppBundle\Services\FunGenerator
        arguments:
            arg2: 'foo_bar'
```

This has the same error as above:

> Missing argument 2 for AppBundle\Services\FunGenerator::__construct() ... (from the cached container)

If we can (from a BC perspective), I think we should fail clearly if we have a string key that does not start with a `$`:

> Unable to resolve service "fun_generator": To specify a named argument, prefix the argument with "$" (e.g. change "arg2" to "$arg2").

Cheers!

**Review by weaverryan — CHANGES_REQUESTED** [review]

**Comment by dunglas** [comment]
> This is tricky... because arg1 might still be autowired (if it had a type-hint)... or... in theory... filled in by some other compiler pass. But allowing $arg2 to be ultimately passed as $arg1 is just wrong.

IMO, this "bug" is already in the code base. The same situation can occurs if you indexed keys like:

```
services:
    fun_generator:
        class: AppBundle\Services\FunGenerator
        arguments:
            1: 'foo_bar'
```

I propose to fix this in another PR (if doable).

> If we can (from a BC perspective), I think we should fail clearly if we have a string key that does not start with a $:

Some people was already using string keys for clarity (even if it wasn't documented as a supported behavior). It's one of the arguments in favor of the `$ ` prefix. If we add this error, it will break some existing app. It's not a big deal because developers was  using undocumented and unsupported behaviors, but is it worth it?

**Comment by weaverryan** [comment]
> IMO, this "bug" is already in the code base. I propose to fix this in another PR (if doable).

I'm very interested in this - since we're really starting to use the `arguments` in more advanced ways, I'd like to be "tighter" on them to keep errors clear. In that other PR, we could propose some final check to make sure all arguments are indexed, and no keys are missing (and throw an exception if there is a problem). And probably more likely (for BC), we could deprecate this loop-hole and first, and then throw a good exception in 4.0 (this means we would not change behavior in 3.x, but could eventually have a really great exception in 4.0).

> Some people was already using string keys for clarity (even if it wasn't documented as a supported behavior).

Other than it being possible to use string keys before, was there any known case (especially public case - e.g. some library, bundle or blog post) where this is happening? Otherwise, my vote would be to throw an exception now, but we could also deprecate its usage in the same way as the above. Definitely, let's do one of them... so 4.0 errors are awesome!

Thanks!

**Comment by iltar** [comment]
>> Some people was already using string keys for clarity (even if it wasn't documented as a supported behavior).

> Other than it being possible to use string keys before, was there any known case (especially public case - e.g. some library, bundle or blog post) where this is happening? Otherwise, my vote would be to throw an exception now, but we could also deprecate its usage in the same way as the above. Definitely, let's do one of them... so 4.0 errors are awesome!

This is something I sometimes see coming by, but very rarely though.

**Comment by nicolas-grekas** [comment]
@weaverryan I juste added a [second commit](https://github.com/symfony/symfony/pull/21383/commits/ca7e42114c5429a37b10d6c8f65dc80eb06675fc) that deprecates using string keys that do not start with a `$`.
@dunglas this commit also fixes some things related to ChildDefinition handling.

I think we can now safely add an `CheckArgumentsValidityPass` that throws an exception if
`array_keys($args) !== range(0, count($args) - 1)` - both for constructor and method args.
I see no reasonable reason where this could make any existing code fail - but with this new feature I see one like @weaverryan spotted - thus the link.

In this PR would be great because that's the one that makes this safeguard required now.

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+                foreach ($parameters as $j => $p) {
+                    if ($key === '$'.$p->name) {
+                        unset($arguments[$key]);
+                        $arguments[$j] = $argument;
+
+                        continue 2;
+                    }
+                }
+
+                throw new InvalidArgumentException(sprintf('Unable to resolve service "%s": method "%s::%s" has no argument name "%s".', $this->currentId, $class, $method, $key));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php]: type: nameD

**Comment by dunglas** [comment]
Comments fixed and added a new `CheckArgumentsValidityPass` (this new pass deserves probably a new PR because it's not related with the current one, but it can be nice to merge everything all together while keeping the 3 commits for clarity).

**Review by nicolas-grekas — APPROVED** [review]
:+1: for me - @weaverryan does this address your comments?

File: src/Symfony/Component/DependencyInjection/Compiler/CheckArgumentsValidityPass.php [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+     */
+    protected function processValue($value, $isRoot = false)
+    {
+        if (!$value instanceof Definition) {
+            return parent::processValue($value, $isRoot);
+        }
+
+        $arguments = $value->getArguments();
+        if ($arguments !== array_values($arguments)) {
+            throw new RuntimeException(sprintf('Arguments\'s keys of the "%s" service must be integers or names of constructor\'s parameters prefixed by the "$" symbol. Check your service definition.', $this->currentId));
```
Comment (weaverryan) [inline src/Symfony/Component/DependencyInjection/Compiler/CheckArgumentsValidityPass.php]: Ok, a few things about this :)

**1**

Take this setup (this is my setup from part A here https://github.com/symfony/symfony/pull/21383#issuecomment-275737454)

```php
class FunGenerator
{
    public function __construct($arg1, $arg2)
    {
    }
}
```

```yml
services:
    fun_generator:
        class: AppBundle\Services\FunGenerator
        arguments:
            $arg2: 'foo_bar'
```

In this case, you WILL hit the above exception. Yay! That's great, because I think this code was added to give the user a better exception with this mistake (which is what I was hoping for!). But, the exception is quote confusing:

> Arguments's keys of the "fun_generator" service must be integers or names of constructor's parameters prefixed by the "$" symbol. Check your service definition.

What happened is, `ResolveNamedArgumentsPass` correctly maps `$arg2` as the `[1]` index... but index `[0]` is missing. So, it hits this exception... but the message is obviously not quite right. This relates to my next points.

**2**
Can we actually get to this point in the code with a key that is still prefixed with $? I think the `ResolveNamedArgumentPass` will already have converted all `$` args to an index arg (or thrown an exception if there is no matching arg name).

**3**

Also, can you even get here by using a non-integer key (which is now deprecated)? The `ResolveNamedArgumentPass` (right after the deprecation trigger) is already converting all non-integer, non-$-prefix arguments to their index/integer value. So, by the time we get here, I don't think it's possible to have *anything* except for integer keys. We're already "correcting" them earlier, which means that I don't think this block is doing anything.


tl;dr; Exactly what situations is this pass meant to catch? Let's make sure our message explains the correct situation.

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+            $method = $parameterBag->resolveValue($method);
+            $parameters = null;
+            $resolvedArguments = array();
+
+            foreach ($arguments as $key => $argument) {
+                if (is_int($key) || '' === $key || '$' !== $key[0]) {
+                    if (!is_int($key)) {
+                        @trigger_error(sprintf('Using key "%s" for defining arguments of method "%s" for service "%s" is deprecated since Symfony 3.3 and will throw an exception in 4.0. Use numeric keys or $named arguments instead.', $key, $method, $this->currentId), E_USER_DEPRECATED);
+                    }
+                    $resolvedArguments[] = $argument;
```
Comment (weaverryan) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php]: This is related to my comment in `CheckArgumentsValidityPass`, point **2** about all non-integer, non-$-prefix values already being changed to index values.

**Review by weaverryan — CHANGES_REQUESTED** [review]

**Comment by weaverryan** [comment]
Love the progress :). I just tried to break everything... again... and have some additional comments. Thanks - this is really tricky... but if we don't get the exceptions & edge cases just right, I'm worried it'll sour the feature.

File: src/Symfony/Component/DependencyInjection/Compiler/CheckArgumentsValidityPass.php (hunk 8 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/CheckArgumentsValidityPass.php]: Exactly the situation where numbering has a hole. The message makes the assumption that if we got there, it might be because of a named arg, even if be don't really know anymore as you spotted.
Which message would be best do you think ?

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/CheckArgumentsValidityPass.php]: Maybe just: arg #1 is missing kind of ?

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+        foreach ($calls as $i => $call) {
+            list($method, $arguments) = $call;
+            $method = $parameterBag->resolveValue($method);
+            $parameters = null;
+            $resolvedArguments = array();
+
+            foreach ($arguments as $key => $argument) {
+                if (is_int($key) || '' === $key || '$' !== $key[0]) {
+                    if (!is_int($key)) {
+                        @trigger_error(sprintf('Using key "%s" for defining arguments of method "%s" for service "%s" is deprecated since Symfony 3.3 and will throw an exception in 4.0. Use numeric keys or $named arguments instead.', $key, $method, $this->currentId), E_USER_DEPRECATED);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php]: False positive you mean ? Aren't index_0 style indexes all resolved in ResolveDefinitionTemplatePass?

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php]: OK, that's a legitimate deprecation: `index="0"` should not be set on this definition:
https://github.com/symfony/monolog-bundle/blob/master/Resources/config/monolog.xml#L35

**Comment by dunglas** [comment]
@weaverryan: should be all good now :) Can you take a look?

File: src/Symfony/Component/DependencyInjection/Compiler/CheckArgumentsValidityPass.php [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+        }
+
+        $i = 0;
+        foreach ($value->getArguments() as $k => $v) {
+            if ($k !== $i++) {
+                if (!is_int($k)) {
+                    throw new RuntimeException(sprintf('Invalid constructor argument for service "%s": integer expected but found string "%s". Check your service definition.', $this->currentId, $k));
+                }
+
+                throw new RuntimeException(sprintf('Invalid constructor argument %d for service "%s": argument %d must be defined before. Check your service definition.', 1 + $k, $this->currentId, $i));
```
Comment (weaverryan) [inline src/Symfony/Component/DependencyInjection/Compiler/CheckArgumentsValidityPass.php]: This issue is triggered because we see argument 2, and that makes us notice that arg 1 is missing. But really, the problem is *entirely* that argument 1 is missing. So, mentioning argument 2 in the error might be counter-productive. What about:

> Missing constructor argument for service "%s": it looks like argument %d was never defined. Check your service definition.'

... and it would be even more awesome if we mentioned the argument name:

> Missing constructor argument for service "%s": it looks like argument %d ($fooBar) was never defined. Check your service definition.'

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+
+                foreach ($parameters as $j => $p) {
+                    if ($key === '$'.$p->name) {
+                        $resolvedArguments[$j] = $argument;
+
+                        continue 2;
+                    }
+                }
+
+                throw new InvalidArgumentException(sprintf('Unable to resolve service "%s": method "%s::%s" has no argument named "%s".', $this->currentId, $class, $method, $key));
```
Comment (weaverryan) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php]: You'll hit this when you mistakenly have, for example, `$arg3: foo` in your services.yml, but no `$arg3` in your constructor/method. So, how about pointing them to their service definition:

> Unable to resolve service "%s": method "%s::%s" has no argument named "%s". Check your service definition.

**Review by weaverryan — CHANGES_REQUESTED** [review]

**Comment by weaverryan** [comment]
Yep, works awesome now! My comments are just error message related. Well, except for this one, but this is "already" a bug and would not work in the current version of things:

```php
class FunGenerator
{
    public function __construct($arg1 = 10, $arg2)
    {
        $this->arg1 = $arg1;
        $this->arg2 = $arg2;
    }
}
```

```yml
services:
    fun_generator:
        class: AppBundle\Services\FunGenerator
        arguments:
            $arg2: 'bar'
```

In 3.2 (if you replaced `$arg2` with `1`), this would fail because the dumped container would incorrectly place 'bar'` as the first argument. After this PR, we'll hit an error in `CheckArgumentsValidityPass`, as it's not smart enough to realize that we *could* read the default value and set it as the argument 1. It's a slight edge case, but not ridiculous, and just not something we ever needed to worry about before we started allowing specific args to be specified.

File: src/Symfony/Component/DependencyInjection/Compiler/CheckArgumentsValidityPass.php [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+        if (!$value instanceof Definition) {
+            return parent::processValue($value, $isRoot);
+        }
+
+        $i = 0;
+        foreach ($value->getArguments() as $k => $v) {
+            if ($k !== $i++) {
+                if (!is_int($k)) {
+                    throw new RuntimeException(sprintf('Invalid constructor argument for service "%s": integer expected but found string "%s". Check your service definition.', $this->currentId, $k));
+                }
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/CheckArgumentsValidityPass.php]: True, at least not when using PassConfig - still possible if the pass is used standalone, if anyone does that.

File: src/Symfony/Component/DependencyInjection/Compiler/CheckArgumentsValidityPass.php (hunk 11 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/CheckArgumentsValidityPass.php]: Giving the name of the argument is possible, but will add significant complexity in the code. Not sure it's worth it.

**Comment by nicolas-grekas** [comment]
> inCheckArgumentsValidityPass`, as it's not smart enough to realize that we could read the default value and set it as the argument

That'd be too smart to me. Requires dealing with reflection, which always adds complexity. I'd prefer consider this an unsupported edge case - the very few needing this will find their way around :)

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php (hunk 12 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php]: LGTM

**Comment by mvrhov** [comment]
@weaverryan: Well that's a beginners mistake I'd say. Parameters with default values should be placed at the end.

**Comment by weaverryan** [comment]
@nicolas-grekas yea, I agree - it's actually a decent (and improved) situation now because the user receives a clear error, instead of a dumped container with the second arg placed in the first position.

@mvrhov it's not a beginner mistake, if both arguments are optional :). Still, not something we should "fix".

File: src/Symfony/Component/DependencyInjection/Compiler/CheckArgumentsValidityPass.php (hunk 11 above)
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Compiler/CheckArgumentsValidityPass.php]: The actual message looks good to me because the order matter. It makes it easier to understand the problem.

File: src/Symfony/Component/DependencyInjection/Compiler/CheckArgumentsValidityPass.php (hunk 13 above)
Comment (weaverryan) [inline src/Symfony/Component/DependencyInjection/Compiler/CheckArgumentsValidityPass.php]: It seems like with autowiring and named args, we've really moved to almost requiring PassConfig to get a working container :). I'm not sure that's a problem - just pointing that out.

**Review by weaverryan — APPROVED** [review]

**Comment by weaverryan** [comment]
👍

**Comment by fabpot** [comment]
@weaverryan Can we close #21376?

**Comment by fabpot** [comment]
@dunglas Looks like another one where a doc PR is needed :)

**Comment by fabpot** [comment]
Thank you @dunglas.

**Comment by weaverryan** [comment]
I've at least got a docs issue opened for this! https://github.com/symfony/symfony-docs/issues/7482

... and I closed my other PR - #21376

**Comment by teohhanhui** [comment]
The `$` prefix is unfortunate. PHP-ism seeping into YAML.

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+        return parent::processValue($value, $isRoot);
+    }
+
+    /**
+     * @param string|null $class
+     * @param string      $method
+     *
+     * @throws InvalidArgumentException
+     *
+     * @return array
```
Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php]: Should be hinted as `\ReflectionParameter[]` so the following calls make more sense

File: src/Symfony/Component/DependencyInjection/ChildDefinition.php [hunk 15]
```diff
@@ -213,12 +213,14 @@ public function getArgument($index)
      */
     public function replaceArgument($index, $value)
```
Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/ChildDefinition.php]: the phpdoc of $index needs to be updated to allow string

Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/ChildDefinition.php]: also ChildDefinition::getArgument doesn't work with string yet. whereas you changed the parent Definition::getArgument.

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php (hunk 14 above)
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php]: This method has been removed during a subsequent refactoring.

**Comment by alcohol** [comment]
For reasons unknown to me, my currently client had tons of service definitions that looked like:

```yaml
service.id:
  class: Foo
  arguments:
    foo: something
    bar: somethingelse
```

All of those broke with the introduction of these changes in 3.3. Was this intentional? Was their use-case undocumented and thus not supported? Cause it did work exactly as expected before 3.3...

**Comment by iltar** [comment]
Afaik, the keys were unused and therefore treated by sequence. Writing it down like this wasn't a supported feature either.

**Comment by alcohol** [comment]
That's what I figured. They were using the keys themselves though in some cases, by retrieving the arguments of the service definition, in some CompilerPass related stuff. Took me a bit to sort everything out 😅
